src/voter_set.rs

```diff
@@ -12,148 +12,241 @@
+				// for weight overflow (not just in debug mode). The protocol
+				// should never run with such voter sets.
+				total_weight = total_weight.checked_add(weight).expect("Voter weight overflow");
+				match voters.entry(id) {
+					Entry::Vacant(e) => {
+						e.insert(VoterInfo {
+							position: 0, // The total order is determined afterwards.
+							weight: VoterWeight(w)
+						});
+					},
+					Entry::Occupied(mut e) => {
+						let v = e.get_mut();
+						let n = v.weight.get() + weight;
+						let w = NonZeroU64::new(n).expect("nonzero + nonzero is nonzero");
+						v.weight = VoterWeight(w);
+					}
+				}
+			}
+		}
 
-	/// Get the length of the set.
-	pub fn len(&self) -> usize { self.voters.len() }
+		if voters.is_empty() {
+			// No non-zero weights; the set would be empty.
+			return None
+		}
 
-	/// Whether the set contains the key.
-	pub fn contains_key(&self, id: &Id) -> bool {
-		self.weights.contains_key(id)
-	}
+		let total_weight = VoterWeight::new(total_weight).expect("voters nonempty; qed");
 
-	// Get voter by index.
-	pub fn voter_by_index(&self, idx: usize) -> &(Id, u64) {
-		&self.voters[idx]
-	}
+		// Establish the total order on the set based on the voter IDs.
+		let mut order = voters.keys().cloned().collect::<Vec<_>>();
+		order.sort_unstable();
+		for (i, id) in order.iter().enumerate() {
+			voters.get_mut(id).expect("def. of order; qed").position = i;
+		}
 
-	/// Get voter info by index.
-	pub fn weight_by_index(&self, idx: usize) -> Option<u64> {
-		self.voters.get(idx).map(|&(_, weight)| weight)
-	}
+		let threshold = threshold(total_weight);
 
-	/// Get the threshold weight.
-	pub fn threshold(&self) -> u64 { self.threshold }
+		Some(VoterSet { voters, order, total_weight, threshold })
+	}
 
-	/// Get the total weight.
-	pub fn total_weight(&self) -> u64 {
-		self.voters.iter().map(|&(_, weight)| weight).sum()
+	/// Get the voter info for the voter with the given ID, if any.
+	pub fn get(&self, id: &Id) -> Option<&VoterInfo> {
+		self.voters.get(id)
 	}
 
-	/// Get the voters.
-	pub fn voters(&self) -> &[(Id, u64)] {
-		&self.voters
+	/// Get the size of the set.
+	pub fn len(&self) -> NonZeroUsize {
+		unsafe {
+			// SAFETY: By VoterSet::new()
+			NonZeroUsize::new_unchecked(self.order.len())
+		}
 	}
-}
 
-impl<Id: Hash + Eq + Clone + Ord> std::iter::FromIterator<(Id, u64)> for VoterSet<Id> {
-	fn from_iter<I: IntoIterator<Item = (Id, u64)>>(iterable: I) -> Self {
-		let iter = iterable.into_iter();
-		let (lower, _) = iter.size_hint();
+	/// Whether the set contains a voter with the given ID.
+	pub fn contains(&self, id: &Id) -> bool {
+		self.voters.contains_key(id)
+	}
 
-		let mut voters = Vec::with_capacity(lower);
-		let mut weights = HashMap::with_capacity(lower);
+	/// Get the nth voter in the set, modulo the size of the set,
+	/// as per the associated total order.
+	pub fn nth_mod(&self, n: usize) -> (&Id, &VoterInfo) {
+		self.nth(n % self.order.len()).unwrap()
```

Comment:
`unwrap`s are definitely not allowed

Reply:
Missed that one, thanks, it was not intentional.